### ISSUE TITLE
API to specify forward mode gradients

### ISSUE THREAD

**Issue by FedericoV** [blackjax-devs/blackjax#201]
Hi,

Thanks for all the work into writing Blackjax: the API is really elegant, and the code is very easy to follow.

I had a quick request: currently, the main API entry point of Blackjax expects the user to pass a (jax) function that computes the log posterior probability.  This function is then internally differentiated using jax primitives, to obtain the function as well as its gradient.

I have a function where the gradients computed using forward mode autodiff are close to 10x faster than those computed with reverse mode.  Unfortunately, there's no easy way to swap between forward and reverse mode autodiff without writing your own integrators (which is not that bad, just, not ideal).

There's a few different ways such an API could go: one option would be to ask users to optionally provide their own derivative, and only calculate the derivative using jax.grad if it's not provided (this seems like the most obvious way forward to me).  Another way would be to give users a choice - and then calculate the derivative internally using `jax.jacfwd` if the user requests forward mode autodiff.

**Comment by junpenglao** [comment]
Either approach sounds fine to me. I slightly prefer optional user specify gradient fun, because it applies to wider user cases (including using `jax.jacfwd` and other analytic gradient implementation)

**Comment by rlouf** [comment]
The first option seems to be the best way forward since, as @junpenglao said, this also allows analytically computed gradient functions to be used.

I am generally reluctant to implement anything that complicates the API, but you're the second person to ask for this possibility, and it will lead to a more uniform API when stochastic gradient MCMC is added to the library.

**Comment by FedericoV** [comment]
I started working on this issue.  Will open a pull request with where I am right now.

**Comment by rlouf** [comment]
Hey @FedericoV, we're deprecating this since you can register custom gradients for your functions. May I ask what your original example was? We're writing a tutorial to explain users how to do this and need an example.

**Comment by FedericoV** [comment]
Hi Remi,

The example we had was a likelihood that included the solution of stiff
ODEs in its computation.  Reverse mode gradients were much slower than
forward mode ones.
